usr/src/uts/common/fs/zfs/dmu.c

```diff
@@ -739,9 +758,29 @@ dmu_free_long_range_impl(objset_t *os, dnode_t *dn, uint64_t offset,
 		ASSERT3U(chunk_begin, >=, offset);
 		ASSERT3U(chunk_begin, <=, chunk_end);
 
-		dmu_tx_t *tx = dmu_tx_create(os);
-		dmu_tx_hold_free(tx, dn->dn_object,
-		    chunk_begin, chunk_end - chunk_begin);
+		chunk_len = chunk_end - chunk_begin;
+
+		mutex_enter(&dp->dp_lock);
+		for (int t = 0; t < TXG_SIZE; t++) {
+			long_free_dirty_all_txgs +=
+			    dp->dp_long_free_dirty_pertxg[t];
+		}
+		mutex_exit(&dp->dp_lock);
+		free_dirty_pct = 100 * long_free_dirty_all_txgs /
+		    zfs_dirty_data_max;
+		/*
+		 * To avoid filling up a TXG with just frees wait for
+		 * the next TXG to open before freeing more chunks if
+		 * we have reached the threshold of frees
+		 */
+		if (dirty_frees_threshold != 0 &&
+		    free_dirty_pct >= dirty_frees_threshold) {
+			txg_wait_open(dp, dp->dp_tx.tx_open_txg + 1);
```

Comment:
It would be cleaner to pass `0` as the 2nd arg -- it does the same thing but with less leaking of txg implementation details and no race condition (although the race wouldn't hurt much).